mirgecom/flux.py

```diff
@@ -35,179 +40,224 @@
+    f = actx.np.where(f_check_plus, f_plus_normal, f)
 
-    The Lax-Friedrichs/Rusanov flux is calculated as:
+    return f
 
-    .. math::
+# }}} low-level flux interfaces
 
-        f_{\mathtt{LFR}} = \frac{1}{2}(\mathbf{F}(q^-) + \mathbf{F}(q^+))
-        + \frac{\lambda}{2}(q^{-} - q^{+})\hat{\mathbf{n}},
 
-    where $q^-, q^+$ are the scalar solution components on the interior and the
-    exterior of the face on which the LFR flux is to be calculated, $\mathbf{F}$ is
-    the vector flux function, $\hat{\mathbf{n}}$ is the face normal, and $\lambda$
-    is the user-supplied jump term coefficient.
+# {{{ Tracepair flux interfaces for operators
 
-    The $\lambda$ parameter is system-specific. Specifically, for the Rusanov flux
-    it is the max eigenvalue of the flux jacobian:
+
+def gradient_flux(u_tpair, normal, beta=0):
+    r"""Compute a numerical flux for the gradient operator.
+
+    The dissipative central gradient flux, $\mathbf{h}$, of a scalar quantity
+    $u$ is calculated as:
 
     .. math::
-        \lambda = \text{max}\left(|\mathbb{J}_{F}(q^-)|,|\mathbb{J}_{F}(q^+)|\right)
 
-    Here, $\lambda$ is a function parameter, leaving the responsibility for the
-    calculation of the eigenvalues of the system-dependent flux Jacobian to the
-    caller.
+        \mathbf{h}({u}^-, {u}^+; \mathbf{n}) = \frac{1}{2}
+        \left(({u}^{+}+{u}^{-}) - \beta(u^{+} - u^{-})\right)\mathbf{\hat{n}}
+
+    where ${u}^-, {u}^+$, are the scalar function values on the interior
+    and exterior of the face on which the central flux is to be calculated, and
+    $\mathbf{\hat{n}}$ is the *normal* vector. Numerical dissipation is added
+    by the solution penalized by a factor $\beta$.
+
+    *u_tpair* is the :class:`~grudge.trace_pair.TracePair` representing the scalar
+    quantities ${u}^-, {u}^+$. *u_tpair* may also represent a vector-quantity
+    :class:`~grudge.trace_pair.TracePair`, and in this case the central scalar flux
+    is computed on each component of the vector quantity as an independent scalar.
 
     Parameters
     ----------
-    cv_tpair: :class:`~grudge.trace_pair.TracePair`
+    u_tpair: :class:`~grudge.trace_pair.TracePair`
+        Trace pair for the face upon which flux calculation is to be performed
+    beta: Number or :class:`~meshmode.dof_array.DOFArray`
+        Optional dissipation strength term
+    normal: numpy.ndarray
+        object array of :class:`~meshmode.dof_array.DOFArray` with outward-pointing
+        normals
+
+    Returns
+    -------
+    numpy.ndarray
+        object array of :class:`~meshmode.dof_array.DOFArray` with the flux for each
+        scalar component.
+    """
+    from arraycontext import outer
+    return outer(u_tpair.avg + .5*beta*u_tpair.diff, normal)
 
-        Solution trace pair for faces for which numerical flux is to be calculated
 
-    f_tpair: :class:`~grudge.trace_pair.TracePair`
+def divergence_flux(trace_pair, normal, alpha=0, beta=0):
```

Comment:
I'm not sure if I like these generic `gradient_flux`/`divergence_flux` functions overall. They seem both overly-general and at the same time not general enough (e.g., what if the quantity being penalized is not the same as the quantity used in the central part?). I wonder if it's better to keep the implementations simple (`gradient_flux_central`, `divergence_flux_central`, etc.), and just add more of them when we need different flux formulations.

Reply:
They are just equation (13) of Mengaldo.